core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -426,39 +426,44 @@ private[spark] class ExecutorAllocationManager(
       }
     }
 
-    override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = synchronized {
+    override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
       val stageId = taskStart.stageId
       val taskId = taskStart.taskInfo.taskId
       val taskIndex = taskStart.taskInfo.index
       val executorId = taskStart.taskInfo.executorId
 
-      // If this is the last pending task, mark the scheduler queue as empty
-      stageIdToTaskIndices.getOrElseUpdate(stageId, new mutable.HashSet[Int]) += taskIndex
-      val numTasksScheduled = stageIdToTaskIndices(stageId).size
-      val numTasksTotal = stageIdToNumTasks.getOrElse(stageId, -1)
-      if (numTasksScheduled == numTasksTotal) {
-        // No more pending tasks for this stage
-        stageIdToNumTasks -= stageId
-        if (stageIdToNumTasks.isEmpty) {
-          allocationManager.onSchedulerQueueEmpty()
+      allocationManager.synchronized {
+        allocationManager.onExecutorAdded(executorId)
```

Comment:
This seems incorrect to me. This marks the executor idle every time it runs a task, which will result in many duplicate executor warnings (see [this line](https://github.com/apache/spark/blob/8d45834debc6986e61831d0d6e982d5528dccc51/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L325)). I think we should do a check instead, and add a huge comment explaining the ordering issue:
```
// This guards against the race condition in which the `SparkListenerTaskStart`
// event is posted after the `SparkListenerBlockManagerAdded` event, which is
// possible because these events are posted in different threads
if (!allocationManager.executorIds.contains(executorId)) {
  allocationManager.onExecutorAdded(executorId)
}
```